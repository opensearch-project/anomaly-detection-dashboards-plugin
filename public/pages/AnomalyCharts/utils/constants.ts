/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

export const INITIAL_ANOMALY_SUMMARY = {
  anomalyOccurrence: 0,
  minAnomalyGrade: 0.0,
  maxAnomalyGrade: 0.0,
  minConfidence: 0.0,
  maxConfidence: 0.0,
  lastAnomalyOccurrence: '',
};

export enum CHART_FIELDS {
  PLOT_TIME = 'plotTime',
  ANOMALY_GRADE = 'anomalyGrade',
  CONFIDENCE = 'confidence',
  DATA = 'data',
  AGG_INTERVAL = 'aggInterval',
  EXPECTED_VALUE = 'expectedValue',
}

export enum CHART_COLORS {
  ANOMALY_GRADE_COLOR = '#D13212',
  FEATURE_DATA_COLOR = '#16191F',
  FEATURE_COLOR = '#fcd529',
  CONFIDENCE_COLOR = '#017F75',
  LIGHT_BACKGROUND = '#FFFFFF',
  DARK_BACKGROUND = '#1D1E24',
}

export const FEATURE_CHART_THEME = [
  {
    lineSeriesStyle: {
      line: {
        strokeWidth: 2,
        visible: true,
        opacity: 0.5,
      },
      point: {
        visible: true,
        stroke: CHART_COLORS.FEATURE_DATA_COLOR,
      },
    },
  },
  {
    colors: {
      vizColors: [CHART_COLORS.FEATURE_DATA_COLOR],
    },
  },
];

export const ANOMALY_CHART_THEME = [
  {
    colors: {
      vizColors: [
        CHART_COLORS.CONFIDENCE_COLOR,
        CHART_COLORS.ANOMALY_GRADE_COLOR,
      ],
    },
  },
];

export const DATE_PICKER_QUICK_OPTIONS = [
  { start: 'now-24h', end: 'now', label: 'last 24 hours' },
  { start: 'now-7d', end: 'now', label: 'last 7 days' },
  { start: 'now-30d', end: 'now', label: 'last 30 days' },
  { start: 'now-90d', end: 'now', label: 'last 90 days' },

  { start: 'now/d', end: 'now', label: 'Today' },
  { start: 'now/w', end: 'now', label: 'Week to date' },
  { start: 'now/M', end: 'now', label: 'Month to date' },
  { start: 'now/y', end: 'now', label: 'Year to date' },
];

export enum LIVE_CHART_CONFIG {
  REFRESH_INTERVAL_IN_SECONDS = 30 * 1000, //poll anomaly result every 30 seconds
  MONITORING_INTERVALS = 60,
}

export const DEFAULT_ANOMALY_SUMMARY = {
  anomalyOccurrence: 0,
  minAnomalyGrade: 0,
  maxAnomalyGrade: 0,
  minConfidence: 0,
  maxConfidence: 0,
  lastAnomalyOccurrence: '-',
};

export const HEATMAP_CHART_Y_AXIS_WIDTH = 30;
