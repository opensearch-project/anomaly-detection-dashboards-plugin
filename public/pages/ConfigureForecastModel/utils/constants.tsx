/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import {
  ModelConfigurationFormikValues,
} from '../../ConfigureForecastModel/models/interfaces';
import { DEFAULT_SHINGLE_SIZE } from '../../../utils/constants';

// FIXME: We intentionally leave some parameters undefined to encourage customers to click 
// the "Suggest parameters" button rather than providing default values like "10 minutes" for interval
export const INITIAL_MODEL_CONFIGURATION_VALUES: ModelConfigurationFormikValues =
  {
    shingleSize: DEFAULT_SHINGLE_SIZE,
    imputationOption: undefined,
    interval: undefined,
    windowDelay: undefined,
    suggestedSeasonality: undefined,
    recencyEmphasis: undefined,
    resultIndex: undefined,
    resultIndexMinAge: 7,
    resultIndexMinSize: 51200,
    resultIndexTtl: 60,
    flattenCustomResultIndex: false,
  };

// an enum for the sparse data handling options
export enum SparseDataOptionValue {
  IGNORE = 'ignore',
  PREVIOUS_VALUE = 'previous_value',
  SET_TO_ZERO = 'set_to_zero',
  CUSTOM_VALUE = 'custom_value',
}
