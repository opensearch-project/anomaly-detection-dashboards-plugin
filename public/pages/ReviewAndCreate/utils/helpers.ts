/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import {
  FEATURE_TYPE,
  FILTER_TYPES,
  UNITS,
  FeatureAttributes,
  Detector,
  UiFeature,
} from '../../../models/interfaces';
import moment from 'moment';
import { get, isEmpty, snakeCase } from 'lodash';
import { DetectorDefinitionFormikValues } from '../../DefineDetector/models/interfaces';
import { FeaturesFormikValues } from '../../ConfigureModel/models/interfaces';
import { CreateDetectorFormikValues } from '../../CreateDetectorSteps/models/interfaces';
import { OPERATORS_QUERY_MAP } from '../../DefineDetector/utils/whereFilters';
import { convertTimestampToNumber } from '../../../utils/utils';
import { CUSTOM_AD_RESULT_INDEX_PREFIX } from '../../../../server/utils/constants';
import { formikToImputationOption, formikToRules } from '../../ConfigureModel/utils/helpers';

export function formikToDetector(values: CreateDetectorFormikValues): Detector {
  const detectionDateRange = values.historical
    ? {
        startTime: convertTimestampToNumber(values.startTime),
        endTime: convertTimestampToNumber(values.endTime),
      }
    : undefined;
  var resultIndex = values.resultIndex;
  if (resultIndex && resultIndex.trim().length > 0) {
    resultIndex = CUSTOM_AD_RESULT_INDEX_PREFIX + resultIndex;
  }
  let detectorBody = {
    name: values.name,
    description: values.description,
    indices: formikToIndices(values.index),
    resultIndex: resultIndex,
    filterQuery: formikToFilterQuery(values),
    uiMetadata: {
      features: { ...featuresToUIMetadata(values.featureList) },
      filters: get(values, 'filters', []),
    },
    featureAttributes: formikToFeatureAttributes(values.featureList),
    timeField: values.timeField,
    detectionInterval: {
      period: { interval: values.interval, unit: UNITS.MINUTES },
    },
    windowDelay: {
      period: { interval: values.windowDelay, unit: UNITS.MINUTES },
    },
    shingleSize: values.shingleSize,
    categoryField: !isEmpty(values?.categoryField)
      ? values.categoryField
      : undefined,
    resultIndexMinAge:
      resultIndex && resultIndex.trim().length > 0
        ? values.resultIndexMinAge
        : undefined,
    resultIndexMinSize:
      resultIndex && resultIndex.trim().length > 0
        ? values.resultIndexMinSize
        : undefined,
    resultIndexTtl:
      resultIndex && resultIndex.trim().length > 0
        ? values.resultIndexTtl
        : undefined,
    flattenCustomResultIndex:
      resultIndex && resultIndex.trim().length > 0
        ? values.flattenCustomResultIndex
        : undefined,
    imputationOption: formikToImputationOption(values.imputationOption),
    rules: formikToRules(values.suppressionRules),
  } as Detector;

  // Optionally add detection date range
  if (detectionDateRange) {
    detectorBody = {
      ...detectorBody,
      //@ts-ignore
      detectionDateRange: detectionDateRange,
    };
  }

  return detectorBody;
}

export const formikToIndices = (indices: { label: string }[]) =>
  indices.map((index) => index.label);

export const formikToFilterQuery = (
  values: CreateDetectorFormikValues | DetectorDefinitionFormikValues
) => {
  let filterQuery = {};
  const filters = get(values, 'filters', []);

  // If we have filters: need to combine into a single filter query.
  // Need to handle each filter type differently when converting
  if (filters.length > 0) {
    let filterArr = [] as any[];
    filters.forEach((filter) => {
      if (filter.filterType === FILTER_TYPES.SIMPLE) {
        filterArr.push(
          //@ts-ignore
          OPERATORS_QUERY_MAP[filter.operator]?.query(filter)
        );
      } else {
        filterArr.push(
          //@ts-ignore
          JSON.parse(filter.query)
        );
      }
    });
    filterQuery = {
      bool: {
        filter: [...filterArr],
      },
    };
  } else {
    filterQuery = { match_all: {} };
  }
  return filterQuery;
};

export function formikToFeatureAttributes(
  values: FeaturesFormikValues[],
  forPreview: boolean = false
): FeatureAttributes[] {
  //@ts-ignore
  return values.map(function (value) {
    const id = forPreview
      ? value.featureId
      : value.newFeature
      ? undefined
      : value.featureId;
    return {
      featureId: id,
      featureName: value.featureName,
      featureEnabled: value.featureEnabled,
      importance: 1,
      aggregationQuery: formikToAggregation(value),
    };
  });
}

export function featuresToUIMetadata(values: FeaturesFormikValues[]) {
  // TODO:: Delete Stale metadata if name is changed
  let features: {
    [key: string]: UiFeature;
  } = {};
  values.forEach((value) => {
    if (value.featureType === FEATURE_TYPE.SIMPLE) {
      features[value.featureName] = {
        featureType: value.featureType,
        aggregationBy: value.aggregationBy,
        aggregationOf:
          value.aggregationOf && value.aggregationOf.length
            ? value.aggregationOf[0].label
            : undefined,
      };
    } else {
      features[value.featureName] = {
        featureType: value.featureType,
      };
    }
  });
  return features;
}

export function formikToAggregation(values: FeaturesFormikValues) {
  if (values.featureType === FEATURE_TYPE.SIMPLE) {
    return values.aggregationBy &&
      values.aggregationOf &&
      values.aggregationOf.length > 0
      ? {
          [snakeCase(values.featureName)]: {
            [values.aggregationBy]: { field: values.aggregationOf[0].label },
          },
        }
      : {};
  }
  return JSON.parse(values.aggregationQuery);
}

export function toStringConfigCell(obj: any): string {
  if (typeof obj != 'undefined') {
    if (obj.hasOwnProperty('period')) {
      let period = obj.period;
      return period.interval + ' ' + period.unit;
    } else if (typeof obj == 'number') {
      // epoch
      return moment(obj).format('MM/DD/YY hh:mm A');
    }
  }
  return '-';
}
