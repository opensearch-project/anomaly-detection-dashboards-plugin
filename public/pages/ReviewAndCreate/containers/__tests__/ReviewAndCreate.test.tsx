/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import React from 'react';
import { Provider } from 'react-redux';
import {
  HashRouter as Router,
  RouteComponentProps,
  Route,
  Switch,
} from 'react-router-dom';
import { render } from '@testing-library/react';
import { ReviewAndCreate } from '../ReviewAndCreate';
import configureStore from '../../../../redux/configureStore';
import { httpClientMock, coreServicesMock } from '../../../../../test/mocks';
import { CoreServicesContext } from '../../../../components/CoreServices/CoreServices';
import { INITIAL_DETECTOR_JOB_VALUES } from '../../../DetectorJobs/utils/constants';
import { INITIAL_MODEL_CONFIGURATION_VALUES } from '../../../ConfigureModel/utils/constants';
import { INITIAL_DETECTOR_DEFINITION_VALUES } from '../../../DefineDetector/utils/constants';

const renderWithRouter = (isEdit: boolean = false) => ({
  ...render(
    <Provider store={configureStore(httpClientMock)}>
      <Router>
        <Switch>
          <Route
            render={(props: RouteComponentProps) => (
              <CoreServicesContext.Provider value={coreServicesMock}>
                <ReviewAndCreate
                  setStep={jest.fn()}
                  values={{
                    ...INITIAL_DETECTOR_DEFINITION_VALUES,
                    ...INITIAL_MODEL_CONFIGURATION_VALUES,
                    ...INITIAL_DETECTOR_JOB_VALUES,
                  }}
                  {...props}
                />
              </CoreServicesContext.Provider>
            )}
          />
        </Switch>
      </Router>
    </Provider>
  ),
});

describe('<ReviewAndCreate /> spec', () => {
  test('renders the component', () => {
    const { container, getByText } = renderWithRouter(false);
    expect(container.firstChild).toMatchSnapshot();
    getByText('Review and create');
    getByText('Detector settings');
    getByText('Model configuration');
    getByText('Detector schedule');
    getByText('Create detector');
  });
});
