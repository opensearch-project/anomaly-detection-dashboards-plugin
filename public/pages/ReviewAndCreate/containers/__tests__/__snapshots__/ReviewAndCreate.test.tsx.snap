// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReviewAndCreate /> spec renders the component 1`] = `
<div
  class="euiPage"
  style="margin-top: -24px;"
>
  <main
    class="euiPageBody"
  >
    <div
      class="euiPageHeader euiPageHeader--responsive"
    >
      <div
        class="euiPageHeaderSection"
      >
        <h1
          class="euiTitle euiTitle--large"
        >
          Review and create 
        </h1>
      </div>
    </div>
    <div
      class="euiPanel euiPanel--paddingMedium"
      style="padding: 20px; margin: 0px;"
    >
      <div
        class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
        style="padding: 0px;"
      >
        <div
          class="euiFlexItem"
        >
          <h3
            class="euiTitle euiTitle--small"
          >
            Detector settings
          </h3>
          <div
            class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
          >
            <div
              class="euiFlexItem content-panel-subTitle"
              style="line-height: normal; max-width: 75%;"
            />
          </div>
        </div>
        <div
          class="euiFlexItem euiFlexItem--flexGrowZero"
        >
          <div
            class="euiFlexGroup euiFlexGroup--gutterMedium euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
          >
            <div
              class="euiFlexItem"
            >
              <button
                class="euiButton euiButton--primary"
                data-test-subj="editDetectorButton"
                type="button"
              >
                <span
                  class="euiButtonContent euiButton__content"
                >
                  <span
                    class="euiButton__text"
                  >
                    Edit
                  </span>
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="euiSpacer euiSpacer--s"
      />
      <div>
        <hr
          class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginSmall"
        />
        <div
          style="padding: 10px 0px;"
        >
          <div
            class="euiFlexGrid euiFlexGrid--gutterLarge euiFlexGrid--wrap euiFlexGrid--responsive"
          >
            <div
              class="euiFlexItem"
            >
              <div
                class="euiFormRow"
                id="random_html_id-row"
                style="width: 250px;"
              >
                <div
                  class="euiFormRow__labelWrapper"
                >
                  <label
                    class="euiFormLabel euiFormRow__label"
                    for="random_html_id"
                  >
                    Name
                  </label>
                </div>
                <div
                  class="euiFormRow__fieldWrapper"
                >
                  <div
                    class="euiText euiText--medium"
                    id="random_html_id"
                  >
                    <p
                      class="enabled"
                    />
                  </div>
                </div>
              </div>
            </div>
            <div
              class="euiFlexItem"
            >
              <div
                class="euiFormRow"
                id="random_html_id-row"
                style="width: 250px;"
              >
                <div
                  class="euiFormRow__labelWrapper"
                >
                  <label
                    class="euiFormLabel euiFormRow__label"
                    for="random_html_id"
                  >
                    Data source index
                  </label>
                </div>
                <div
                  class="euiFormRow__fieldWrapper"
                >
                  <div
                    class="euiText euiText--medium"
                    id="random_html_id"
                  >
                    <p
                      class="enabled"
                    />
                  </div>
                </div>
              </div>
            </div>
            <div
              class="euiFlexItem"
            >
              <div
                class="euiFormRow"
                id="random_html_id-row"
                style="width: 250px;"
              >
                <div
                  class="euiFormRow__labelWrapper"
                >
                  <label
                    class="euiFormLabel euiFormRow__label"
                    for="random_html_id"
                  >
                    Data filter
                  </label>
                </div>
                <div
                  class="euiFormRow__fieldWrapper"
                >
                  <div
                    class="euiText euiText--medium"
                  >
                    <p
                      class="enabled"
                    >
                      -
                    </p>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="euiFlexItem"
            >
              <div
                class="euiFormRow"
                id="random_html_id-row"
                style="width: 250px;"
              >
                <div
                  class="euiFormRow__labelWrapper"
                >
                  <label
                    class="euiFormLabel euiFormRow__label"
                    for="random_html_id"
                  >
                    Detector interval
                  </label>
                </div>
                <div
                  class="euiFormRow__fieldWrapper"
                >
                  <div
                    class="euiText euiText--medium"
                    id="random_html_id"
                  >
                    <p
                      class="enabled"
                    >
                      10 Minutes
                    </p>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="euiFlexItem"
            >
              <div
                class="euiFormRow"
                id="random_html_id-row"
                style="width: 250px;"
              >
                <div
                  class="euiFormRow__labelWrapper"
                >
                  <label
                    class="euiFormLabel euiFormRow__label"
                    for="random_html_id"
                  >
                    Description
                  </label>
                </div>
                <div
                  class="euiFormRow__fieldWrapper"
                >
                  <div
                    class="euiText euiText--medium"
                    id="random_html_id"
                  >
                    <p
                      class="enabled"
                    />
                  </div>
                </div>
              </div>
            </div>
            <div
              class="euiFlexItem"
            >
              <div
                class="euiFormRow"
                id="random_html_id-row"
                style="width: 250px;"
              >
                <div
                  class="euiFormRow__labelWrapper"
                >
                  <label
                    class="euiFormLabel euiFormRow__label"
                    for="random_html_id"
                  >
                    Timestamp
                  </label>
                </div>
                <div
                  class="euiFormRow__fieldWrapper"
                >
                  <div
                    class="euiText euiText--medium"
                    id="random_html_id"
                  >
                    <p
                      class="enabled"
                    />
                  </div>
                </div>
              </div>
            </div>
            <div
              class="euiFlexItem"
            >
              <div
                class="euiFormRow"
                id="random_html_id-row"
                style="width: 250px;"
              >
                <div
                  class="euiFormRow__labelWrapper"
                >
                  <label
                    class="euiFormLabel euiFormRow__label"
                    for="random_html_id"
                  >
                    Window delay
                  </label>
                </div>
                <div
                  class="euiFormRow__fieldWrapper"
                >
                  <div
                    class="euiText euiText--medium"
                    id="random_html_id"
                  >
                    <p
                      class="enabled"
                    >
                      1 Minutes
                    </p>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="euiSpacer euiSpacer--l"
    />
    <div
      class="euiPanel euiPanel--paddingMedium"
      style="padding: 20px;"
    >
      <div
        class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
        style="padding: 0px;"
      >
        <div
          class="euiFlexItem"
        >
          <h3
            class="euiTitle euiTitle--small"
          >
            Model configuration
          </h3>
          <div
            class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
          >
            <div
              class="euiFlexItem content-panel-subTitle"
              style="line-height: normal; max-width: 75%;"
            />
          </div>
        </div>
        <div
          class="euiFlexItem euiFlexItem--flexGrowZero"
        >
          <div
            class="euiFlexGroup euiFlexGroup--gutterMedium euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
          >
            <div
              class="euiFlexItem"
            >
              <button
                class="euiButton euiButton--primary"
                type="button"
              >
                <span
                  class="euiButtonContent euiButton__content"
                >
                  <span
                    class="euiButton__text"
                  >
                    Edit
                  </span>
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="euiSpacer euiSpacer--s"
      />
      <div>
        <hr
          class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginSmall"
        />
        <div
          style="padding: 10px 0px;"
        >
          <div>
            <div
              class="euiBasicTable header-single-value-euiBasicTable"
            >
              <div>
                <div
                  class="euiTableHeaderMobile"
                >
                  <div
                    class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsBaseline euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow"
                  >
                    <div
                      class="euiFlexItem euiFlexItem--flexGrowZero"
                    />
                    <div
                      class="euiFlexItem euiFlexItem--flexGrowZero"
                    />
                  </div>
                </div>
                <table
                  class="euiTable euiTable--responsive"
                  id="random_html_id"
                  tabindex="-1"
                >
                  <caption
                    class="euiScreenReaderOnly euiTableCaption"
                  />
                  <thead>
                    <tr>
                      <th
                        class="euiTableHeaderCell"
                        data-test-subj="tableHeaderCell_categoryField_0"
                        role="columnheader"
                        scope="col"
                      >
                        <div
                          class="euiTableCellContent"
                        >
                          <span
                            class="euiTableCellContent__text"
                            title="Category field"
                          >
                            Category field
                          </span>
                        </div>
                      </th>
                      <th
                        class="euiTableHeaderCell"
                        data-test-subj="tableHeaderCell_windowSize_1"
                        role="columnheader"
                        scope="col"
                      >
                        <div
                          class="euiTableCellContent"
                        >
                          <span
                            class="euiTableCellContent__text"
                            title="Window size"
                          >
                            Window size
                          </span>
                        </div>
                      </th>
                    </tr>
                  </thead>
                  <tbody>
                    <tr
                      class="euiTableRow"
                    >
                      <td
                        class="euiTableRowCell"
                      >
                        <div
                          class="euiTableRowCell__mobileHeader euiTableRowCell--hideForDesktop"
                        >
                          Category field
                        </div>
                        <div
                          class="euiTableCellContent"
                        >
                          <span
                            class="euiTableCellContent__text"
                          >
                            -
                          </span>
                        </div>
                      </td>
                      <td
                        class="euiTableRowCell"
                      >
                        <div
                          class="euiTableRowCell__mobileHeader euiTableRowCell--hideForDesktop"
                        >
                          Window size
                        </div>
                        <div
                          class="euiTableCellContent"
                        >
                          <span
                            class="euiTableCellContent__text"
                          >
                            4
                          </span>
                        </div>
                      </td>
                    </tr>
                  </tbody>
                </table>
              </div>
            </div>
            <div
              class="euiSpacer euiSpacer--l"
            />
            <div
              class="euiPanel euiPanel--paddingMedium"
              style="padding: 20px;"
            >
              <div
                class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
                style="padding: 0px;"
              >
                <div
                  class="euiFlexItem"
                >
                  <div
                    class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow euiFlexGroup--responsive"
                  >
                    <div
                      class="euiFlexItem"
                    >
                      <h3
                        class="euiTitle euiTitle--small"
                        style="display: flex; flex-direction: row;"
                      >
                        <p>
                          Features
                           
                        </p>
                        <p
                          style="color: rgb(83, 89, 102);"
                        >
                          (0)
                        </p>
                      </h3>
                    </div>
                  </div>
                  <div
                    class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
                  >
                    <div
                      class="euiFlexItem content-panel-subTitle"
                      style="line-height: normal; max-width: 75%;"
                    />
                  </div>
                </div>
                <div
                  class="euiFlexItem euiFlexItem--flexGrowZero"
                >
                  <div
                    class="euiFlexGroup euiFlexGroup--gutterMedium euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
                  >
                    <div
                      class="euiFlexItem"
                    />
                  </div>
                </div>
              </div>
              <div>
                <hr
                  class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginSmall"
                />
                <div
                  style="padding: 10px 0px;"
                >
                  <div
                    class="euiBasicTable"
                  >
                    <div>
                      <div
                        class="euiTableHeaderMobile"
                      >
                        <div
                          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsBaseline euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow"
                        >
                          <div
                            class="euiFlexItem euiFlexItem--flexGrowZero"
                          />
                          <div
                            class="euiFlexItem euiFlexItem--flexGrowZero"
                          >
                            <div
                              class="euiTableSortMobile"
                            >
                              <div
                                class="euiPopover euiPopover--anchorDownRight"
                              >
                                <div
                                  class="euiPopover__anchor"
                                >
                                  <button
                                    class="euiButtonEmpty euiButtonEmpty--primary euiButtonEmpty--xSmall euiButtonEmpty--flushRight"
                                    type="button"
                                  >
                                    <span
                                      class="euiButtonContent euiButtonContent--iconRight euiButtonEmpty__content"
                                    >
                                      <svg
                                        aria-hidden="true"
                                        class="euiIcon euiIcon--medium euiIcon-isLoading euiButtonContent__icon"
                                        focusable="false"
                                        height="16"
                                        role="img"
                                        viewBox="0 0 16 16"
                                        width="16"
                                        xmlns="http://www.w3.org/2000/svg"
                                      />
                                      <span
                                        class="euiButtonEmpty__text"
                                      >
                                        Sorting
                                      </span>
                                    </span>
                                  </button>
                                </div>
                              </div>
                            </div>
                          </div>
                        </div>
                      </div>
                      <table
                        class="euiTable euiTable--responsive"
                        id="random_html_id"
                        tabindex="-1"
                      >
                        <caption
                          class="euiScreenReaderOnly euiTableCaption"
                        />
                        <thead>
                          <tr>
                            <th
                              aria-live="polite"
                              aria-sort="ascending"
                              class="euiTableHeaderCell"
                              data-test-subj="tableHeaderCell_name_0"
                              role="columnheader"
                              scope="col"
                            >
                              <button
                                class="euiTableHeaderButton euiTableHeaderButton-isSorted"
                                data-test-subj="tableHeaderSortButton"
                                type="button"
                              >
                                <span
                                  class="euiTableCellContent"
                                >
                                  <span
                                    class="euiTableCellContent__text"
                                    title="Feature name; Sorted in ascending order"
                                  >
                                    Feature name
                                  </span>
                                  <svg
                                    aria-hidden="true"
                                    aria-label="Sorted in ascending order"
                                    class="euiIcon euiIcon--medium euiIcon-isLoading euiTableSortIcon"
                                    focusable="false"
                                    height="16"
                                    role="img"
                                    viewBox="0 0 16 16"
                                    width="16"
                                    xmlns="http://www.w3.org/2000/svg"
                                  />
                                  <span
                                    class="euiScreenReaderOnly"
                                  >
                                    Click to sort in descending order
                                  </span>
                                </span>
                              </button>
                            </th>
                            <th
                              class="euiTableHeaderCell"
                              data-test-subj="tableHeaderCell_definition_1"
                              role="columnheader"
                              scope="col"
                            >
                              <div
                                class="euiTableCellContent"
                              >
                                <span
                                  class="euiTableCellContent__text"
                                  title="Feature definition"
                                >
                                  Feature definition
                                </span>
                              </div>
                            </th>
                            <th
                              class="euiTableHeaderCell"
                              data-test-subj="tableHeaderCell_state_2"
                              role="columnheader"
                              scope="col"
                            >
                              <div
                                class="euiTableCellContent"
                              >
                                <span
                                  class="euiTableCellContent__text"
                                  title="Feature state"
                                >
                                  Feature state
                                </span>
                              </div>
                            </th>
                          </tr>
                        </thead>
                        <tbody>
                          <tr
                            class="euiTableRow"
                          >
                            <td
                              class="euiTableRowCell euiTableRowCell--isMobileFullWidth"
                              colspan="3"
                            >
                              <div
                                class="euiTableCellContent euiTableCellContent--alignCenter"
                              >
                                <span
                                  class="euiTableCellContent__text"
                                >
                                  No items found
                                </span>
                              </div>
                            </td>
                          </tr>
                        </tbody>
                      </table>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="euiSpacer euiSpacer--l"
    />
    <div
      class="euiPanel euiPanel--paddingMedium"
      style="padding: 20px; margin: 0px;"
    >
      <div
        class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
        style="padding: 0px;"
      >
        <div
          class="euiFlexItem"
        >
          <h3
            class="euiTitle euiTitle--small"
          >
            Detector schedule
          </h3>
          <div
            class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
          >
            <div
              class="euiFlexItem content-panel-subTitle"
              style="line-height: normal; max-width: 75%;"
            />
          </div>
        </div>
        <div
          class="euiFlexItem euiFlexItem--flexGrowZero"
        >
          <div
            class="euiFlexGroup euiFlexGroup--gutterMedium euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
          >
            <div
              class="euiFlexItem"
            >
              <button
                class="euiButton euiButton--primary"
                data-test-subj="editScheduleButton"
                type="button"
              >
                <span
                  class="euiButtonContent euiButton__content"
                >
                  <span
                    class="euiButton__text"
                  >
                    Edit
                  </span>
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="euiSpacer euiSpacer--s"
      />
      <div>
        <hr
          class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginSmall"
        />
        <div
          style="padding: 10px 0px;"
        >
          <div
            class="euiFlexGrid euiFlexGrid--gutterLarge euiFlexGrid--fourths euiFlexGrid--responsive"
          >
            <div
              class="euiFlexItem"
            >
              <div
                class="euiFormRow"
                id="random_html_id-row"
                style="width: 250px;"
              >
                <div
                  class="euiFormRow__labelWrapper"
                >
                  <label
                    class="euiFormLabel euiFormRow__label"
                    for="random_html_id"
                  >
                    Real-time detector
                  </label>
                </div>
                <div
                  class="euiFormRow__fieldWrapper"
                >
                  <div
                    class="euiText euiText--medium"
                    id="random_html_id"
                  >
                    <p
                      class="enabled"
                    >
                      Start automatically
                    </p>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="euiFlexItem"
            >
              <div
                class="euiFormRow"
                id="random_html_id-row"
                style="width: 250px;"
              >
                <div
                  class="euiFormRow__labelWrapper"
                >
                  <label
                    class="euiFormLabel euiFormRow__label"
                    for="random_html_id"
                  >
                    Historical analysis
                  </label>
                </div>
                <div
                  class="euiFormRow__fieldWrapper"
                >
                  <div
                    class="euiText euiText--medium"
                    id="random_html_id"
                  >
                    <p
                      class="enabled"
                    >
                      Disabled
                    </p>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </main>
</div>
`;
